Bottom: 5141c5ff1d9ee840850a6e31afaa49f76eb31cc1
Top:    b169d4b2896f7a4b905804db035e21533237e080
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-16 09:17:44 -0500

Populate buildbot with gitlab public keys


---
diff --git a/ansible-playbook/buildbot.yml b/ansible-playbook/buildbot.yml
index 509dbc8d..cfb91cca 100644
--- a/ansible-playbook/buildbot.yml
+++ b/ansible-playbook/buildbot.yml
@@ -6,14 +6,20 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
-  roles:
-  - name: Build and start the Buildbot master container
-    role: buildbot-master
-    tags:
+    - buildbot
     - buildbot-master
+  tasks:
+  - name: Scan Gitlab ssh keys
+    command: ssh-keyscan {{gitlab_hostname}}
+    register: gitlab_keys
+  - name: Build and start the Buildbot master container
+    import_role: 
+      name: buildbot-master
+    vars:
+      known_hosts: "{{gitlab_keys.stdout}}"
   - name: Add buildbot user to Gitlab
-    role: gitlab-user
+    import_role: 
+      name: gitlab-user
     vars:
       new_user: buildbot
       new_user_name: Build Robot
@@ -30,7 +36,11 @@
     - buildbot-worker
   vars:
     buildbot_master: "{{ hostvars['buildbot-master'].ip_addr }}"
+    known_hosts: "{{gitlab_keys.stdout}}"
   tasks:
+  - name: Scan Gitlab ssh keys
+    command: ssh-keyscan {{gitlab_hostname}}
+    register: gitlab_keys
   - import_role: 
       name: buildbot-worker-ntel
   - import_role: 
@@ -42,13 +52,13 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
+    - buildbot
   roles:
   - name: Add buildbot_master pubkey to buildbot user
     role: gitlab-sshkey
     vars:
       username: buildbot
-      pubkey: "{{ hostvars.buildbot_master.pubkey }}"
+      pubkey: "{{ hostvars['buildbot-master'].pubkey }}"
       title: "buildbot_master"
   - name: Add buildbot_worker_ntel pubkey to buildbot user
     role: gitlab-sshkey
diff --git a/docker/buildbot-master/buildbot/start.sh b/docker/buildbot-master/buildbot/start.sh
index bf4c4e5a..7813d128 100755
--- a/docker/buildbot-master/buildbot/start.sh
+++ b/docker/buildbot-master/buildbot/start.sh
@@ -6,5 +6,13 @@ SSH_FILE=${SSH_FILE-"$HOME/.ssh/id_rsa"}
 
 ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/null 2>/dev/null
 
+if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
+then
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
+fi
+
 # Run original buildbot entrypoint
 exec "/usr/src/buildbot/docker/start_buildbot.sh"
diff --git a/docker/buildbot-worker-ntel/buildbot/start.sh b/docker/buildbot-worker-ntel/buildbot/start.sh
index f025e661..e7d817e3 100755
--- a/docker/buildbot-worker-ntel/buildbot/start.sh
+++ b/docker/buildbot-worker-ntel/buildbot/start.sh
@@ -10,8 +10,10 @@ ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/nul
 
 if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
 then
-    mkdir -p "${HOME}/.ssh"
-    cp -v "$BUILDBOT_KNOWN_HOSTS_FILE" "${HOME}/.ssh/known_hosts"
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
 fi
 
 ln -v -s -f ${HOME}/buildbot.tac $BUILDBOT_DATA
diff --git a/docker/buildbot-worker-ptxdist/buildbot/start.sh b/docker/buildbot-worker-ptxdist/buildbot/start.sh
index f025e661..e7d817e3 100755
--- a/docker/buildbot-worker-ptxdist/buildbot/start.sh
+++ b/docker/buildbot-worker-ptxdist/buildbot/start.sh
@@ -10,8 +10,10 @@ ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/nul
 
 if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
 then
-    mkdir -p "${HOME}/.ssh"
-    cp -v "$BUILDBOT_KNOWN_HOSTS_FILE" "${HOME}/.ssh/known_hosts"
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
 fi
 
 ln -v -s -f ${HOME}/buildbot.tac $BUILDBOT_DATA
diff --git a/roles/buildbot-master/defaults/main.yml b/roles/buildbot-master/defaults/main.yml
index a4296e57..3e81b9b2 100644
--- a/roles/buildbot-master/defaults/main.yml
+++ b/roles/buildbot-master/defaults/main.yml
@@ -33,6 +33,7 @@ config_hostdir: '{{ docker_restore_config_base_dir }}/{{ container_name }}'
 # environment passed to the container
 container_env:
   BUILDBOT_DATABASE: 'sqlite:///{{ data_path }}/state.sqlite'
+  BUILDBOT_KNOWN_HOSTS_FILE: '{{secrets_path}}/{{secrets_known_hosts}}'
   BUILDBOT_SMTP_HOST: 'relay.dnsexit.com'
   BUILDBOT_SMTP_USER: 'Dean.cox'
   BUILDBOT_SMTP_PASS: 'no5156'
@@ -72,6 +73,7 @@ container_volumes:
   - '{{ config_hostdir }}/master.cfg:{{ data_path }}/master.cfg'
   - '{{ config_hostdir }}/openembedded.py:{{ data_path }}/openembedded.py'
   - '{{ config_hostdir }}/ptxdist.py:{{ data_path }}/ptxdist.py'
+  - '{{ secrets_hostdir }}/{{secrets_known_hosts}}:{{ secrets_path }}/{{secrets_known_hosts}}'
 
 # mount path of the data volume
 data_path: '/buildbot'
diff --git a/roles/buildbot-master/tasks/main.yml b/roles/buildbot-master/tasks/main.yml
index 1c0c1eeb..f2d530d1 100644
--- a/roles/buildbot-master/tasks/main.yml
+++ b/roles/buildbot-master/tasks/main.yml
@@ -31,6 +31,16 @@
   loop:
     - buildbot.tac.j2
 
+- name: Create secrets volume
+  file:
+    path: '{{ secrets_hostdir }}'
+    state: directory
+
+- name: Populate known_hosts secret
+  copy:
+    content: "{{ known_hosts }}\n"
+    dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
+
 - name: Checkout image repo
   git:
     repo: '{{ image_repo }}'
diff --git a/roles/buildbot-worker-ntel/tasks/main.yml b/roles/buildbot-worker-ntel/tasks/main.yml
index 140f3e6b..b1bcf349 100644
--- a/roles/buildbot-worker-ntel/tasks/main.yml
+++ b/roles/buildbot-worker-ntel/tasks/main.yml
@@ -50,6 +50,16 @@
         chown -c -R {{buildbot_uid}}:{{buildbot_uid}} {{item}};
       fi"
 
+- name: Create secrets volume
+  file:
+    path: '{{ secrets_hostdir }}'
+    state: directory
+
+- name: Populate known_hosts secret
+  copy:
+    content: "{{ known_hosts }}\n"
+    dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
+
 - name: Checkout image repo
   git:
     repo: '{{ image_repo }}'
diff --git a/roles/buildbot-worker-ptxdist/tasks/main.yml b/roles/buildbot-worker-ptxdist/tasks/main.yml
index 323363a8..f24bb829 100644
--- a/roles/buildbot-worker-ptxdist/tasks/main.yml
+++ b/roles/buildbot-worker-ptxdist/tasks/main.yml
@@ -50,6 +50,16 @@
         chown -c -R {{buildbot_uid}}:{{buildbot_uid}} {{item}};
       fi"
 
+- name: Create secrets volume
+  file:
+    path: '{{ secrets_hostdir }}'
+    state: directory
+
+- name: Populate known_hosts secret
+  copy:
+    content: "{{ known_hosts }}\n"
+    dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
+
 - name: Checkout image repo
   git:
     repo: '{{ image_repo }}'
